pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -316,6 +317,229 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
 	p.cpusToReuse[string(pod.UID)] = p.cpusToReuse[string(pod.UID)].Difference(cset)
 }
 
+// validatePodScopeResources checks for the "empty shared pool" scenario. This occurs
+// when the sum of exclusive container CPU requests consumes the entire pod-level
+// budget, leaving no CPUs for containers that require a shared pool. Such a
+// configuration is invalid because it would lead to containers in the shared
+// pool having an empty cpuset, causing them to run on the node's shared pool
+// and breaking NUMA affinity.
+func (p *staticPolicy) validatePodScopeResources(logger logr.Logger, pod *v1.Pod) error {
+	podTotalCPUs := p.podGuaranteedCPUs(logger, pod)
+
+	// Check for empty shared pool for standard init containers.
+
+	sumOfSidecarExclusiveCPUs := 0
+	maxStandardInitExclusiveCPUs := 0
+	hasSharedStandardInitContainer := false
+	for _, container := range pod.Spec.InitContainers {
+		guaranteedCPUs := p.guaranteedCPUs(logger, pod, &container)
+		if podutil.IsRestartableInitContainer(&container) {
+			sumOfSidecarExclusiveCPUs += guaranteedCPUs
+			continue
+		}
+
+		if guaranteedCPUs > 0 && guaranteedCPUs > maxStandardInitExclusiveCPUs {
+			maxStandardInitExclusiveCPUs = guaranteedCPUs
+		} else {
+			hasSharedStandardInitContainer = true
+		}
+	}
+
+	// This check ensures that the pool of CPUs available to standard init containers
+	// (the total pod budget minus any sidecars) is large enough to satisfy the
+	// request of the largest standard init container that requires exclusive CPUs.
+	// Since standard init containers run sequentially, we only need to fit the largest one.
+	poolForStandardInits := podTotalCPUs - sumOfSidecarExclusiveCPUs
+	if maxStandardInitExclusiveCPUs > poolForStandardInits {
```

Comment:
I think this logic isn't quite right, since sidecar startup is ordered with init containers. For example, you could have a first init container that takes up all the CPUs, followed by several sidecars. This should be acceptable, because the initial init container will terminate before the sidecars start. I think you should make `maxStandardInitExclusiveCPUs` include the cumulative sidecar request to that point (see comment above).